### PR TITLE
kubeadm: prevent bootstrap of nodes with known names

### LINKED ISSUE #1711: reject a node joining the cluster if a node with the same name already exists
the discussion in https://github.com/kubernetes/kubernetes/issues/79779
revealed that adding multiple nodes with the same name as a control-plane node name can bring the cluster down.

while this can be considered an operator mistake, given k8s does not support nodes with the same name, kubeadm should take action on preventing the bootstrap of such nodes on `kubeadm join` time.

### PR DESCRIPTION
**Pull request by neolit123**
**What this PR does / why we need it**:
If a Node name in the cluster is already taken and this Node is Ready,
prevent TLS bootsrap on "kubeadm join" and exit early.

This change requires that a new ClusterRole is granted to the
"system:bootstrappers:kubeadm:default-node-token" group to be
able get Nodes in the cluster. The same group already has access
to obtain objects such as the KubeletConfiguration and kubeadm's
ClusterConfiguration.

The motivation of this change is to prevent undefined behavior
and the potential control-plane breakdown if such a cluster
is racing to have two nodes with the same name for long periods
of time.

The following values are validated in the following precedence
from lower to higher:
- actual hostname
- NodeRegistration.Name (or "--node-name") from JoinConfiguration
- "--hostname-override" passed via kubeletExtraArgs

If the user decides to not let kubeadm know about a custom node name
and to instead override the hostname from a kubelet systemd unit file,
kubeadm will not be able to detect the problem.

**Which issue(s) this PR fixes**:

Fixes kubernetes/kubeadm#1711

**Special notes for your reviewer**:
NONE

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: reject a node joining the cluster if a node with the same name already exists
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

/kind feature
/priority important-longterm

### CODE REVIEW

File: cmd/kubeadm/app/cmd/phases/join/kubelet.go
Comment (SataQiu) [inline cmd/kubeadm/app/cmd/phases/join/kubelet.go]:
```suggestion
	node, err := bootstrapClient.CoreV1().Nodes().Get(nodeName , metav1.GetOptions{})
```
@neolit123 In my view, we can use `Get` directly instead of `List`. Right?

**Comment by liggitt** [comment]
what are the implications for cleaning up a node to set it up again? does this mean that `kubeadm reset` is no longer sufficient to allow `kubeadm join` to succeed again?

File: cmd/kubeadm/app/cmd/phases/join/kubelet.go
Comment (liggitt) [inline cmd/kubeadm/app/cmd/phases/join/kubelet.go]: should this be a skippable preflight check instead?

File: cmd/kubeadm/app/cmd/phases/join/kubelet.go
Comment (liggitt) [inline cmd/kubeadm/app/cmd/phases/join/kubelet.go]: can the logic be shared? two places doing this fallback are likely to drift

**Review by rosti — COMMENTED** [review]
Thanks @neolit123 !
We may also try to nuke the Node object on reset, but permission wise IDK if this is feasible.

**Comment by liggitt** [comment]
> We may also try to nuke the Node object on reset, but permission wise IDK if this is feasible.

if `kubeadm reset` is run with the API permissions of the kubelet, it will not have the ability to delete the Node API object (I didn't think it made any API calls today)

> I suppose, that Get is slightly more efficient, but IDK that for sure. Also, IDK what's the difference from security standpoint in allowing folks, with bootstrap tokens, to list all nodes vs get nodes.
> I'll leave that on the trusted advice of @liggitt .

Get is way more efficient on large clusters.

A bootstrap token can obtain a node client credential that allows get/list/watch of all nodes, so read-only access to Get a node is not concerning.

File: cmd/kubeadm/app/cmd/phases/join/kubelet.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/join/kubelet.go]: no so sure about this, stale nodes are a subject of the administrator calling `kubectl delete`. `kubeadm join` failing during boostrap would indicate that the cluster requires admin intervetion.

if the user is allowed to skip this check, it can lead to catastrophic outcome - e.g. imaging a case where the operator responsible for the worker join is not the cluster admin.

let's talk more about this during the office hours, possibly next week.

**Comment by neolit123** [comment]
> Get is way more efficient on large clusters.

completely forgot about `get` in the early hours of the day.

> We may also try to nuke the Node object on reset, but permission wise IDK if this is feasible.

possibly a separate PR, but also needs permission evaluation.
my initial reaction is to leave this to the admin group.

File: cmd/kubeadm/app/cmd/phases/join/kubelet.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/join/kubelet.go]: it's slightly more difficult than it seems and will extend the scope of this PR (i.e. include refactors), but i can give it a shot.

**Comment by neolit123** [comment]
> what are the implications for cleaning up a node to set it up again? does this mean that kubeadm reset is no longer sufficient to allow kubeadm join to succeed again?

this is a problem that i though about after sending the PR.

we either have to:
- extend reset to delete nodes.
- modify this check to only fail on existing `Ready` nodes.

**Comment by neolit123** [comment]
@rosti @liggitt @SataQiu 

i have updated this PR:
- using "GET" for the Node object
- use the same common utility for node name precedence between the place where we write kubelet flags and join new nodes
- still allow a new Node to join if the same Node name the cluster exists, but it's status is not Ready. this allows re-join on failure.

**Comment by neolit123** [comment]
/hold
for review

File: cmd/kubeadm/app/cmd/phases/join/kubelet.go
Comment (rosti) [inline cmd/kubeadm/app/cmd/phases/join/kubelet.go]: We may want to dump a message towards the user, that this node cannot be joined because of name collision and some appropriate steps might be necessary (like ensuring the old node is truly gone and kubectl delete gets called).

WDYT?

**Review by rosti — APPROVED** [review]
Thanks @neolit123 !
Implementation wise, it looks good.
We may want to add a more thorough message to the user along with steps how to fix the problem.
/lgtm

**Review by rosti — APPROVED** [review]
Thanks @neolit123 !
/lgtm

**Comment by kvaps** [comment]
Hi, this change caused two new issues: https://github.com/kubernetes/kubernetes/issues/89501, https://github.com/kubernetes/kubernetes/issues/89512

**Pull request merged**
